### PR TITLE
Heapster tutorial

### PR DESCRIPTION
**Pull request by scuellar**
Add a tutorial to get up to speed with Heapster.

### CODE REVIEW

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:63]: Please add a link to that README

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:76]: This should be `cd ../..`

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:337]: Instead of expecting the user to be in the `heapster-saw` directory for all this tutorial, you should have them be in `heapster-saw/examples`, so that the user can access the Makefile listed there. Maybe put an explicit `cd heapster-saw/examples` command in section 1 (or 0)?

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:368]: I'm confused by this comment, because `make linked_list.bc` works for me (when run in the `heapster-saw/examples` directory)

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:384]: Typo: "wiht" -> "with"

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:417]: Typo: "it's" -> "its"

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:416]: You should maybe introduce `heapster_init_env_from_file` first, because almost all of the examples (including specifically the `linked_list` example) use it, and then you can add an extra few sentences mentioning that `heapster_init_env` is a special case

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:481]: You should use the term "Heapster types" instead of "Heapster specifications", because, like you say later, Heapster types aren't full functional specifications, but instead mostly deal with memory shapes and whether LLVM values are pointers or word values.

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:490]: It's a little confusing that you introduce some notation here that is close to but not quite the same as Heapster notation. Maybe, instead of trying to first define the type in an abstract way and refining it in the next sub-section to the Heapster syntax, you could just give the Heapster type for `is_elem` and explain each of its pieces

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:556]: Explain the `rw` argument, because this is the first example that uses it

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:558]: Expand the part about the cases into its own 1-2 sentences describing the two different cases and what their syntax means, even if briefly. For instance, say something like "In this case, a `List64` is either a null pointer, which is expressed with the type `eq(llvmword(0))` that declares that it equal to the word value 0, or it is a pointer where offset 0 points to a value satisfying the `int64<>` type and where offset 8 points to a value satisfying the `List64<rw>` permission, i.e., it is recursively a `List64<rw>`.

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:530]: Mention that uses of named permissions are written like `int64<>`, where the `<>` is the empty list of arguments.

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [reply heapster-saw/doc/tutorial/tutorial.md:558]: Also reiterate that the syntax `List64<rw>` is the result of applying the `List64` named permission to the single argument `rw`.

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [inline heapster-saw/doc/tutorial/tutorial.md:578]: This explanation of every predicate requiring a read-only or read-write modality is not quite right. The reason we write `List64<R>` here is because `List64` was defined to take a single argument of type `rwmodality`. Read-write access to a list would require the type `List64<W>`. You can't use `<>` in place of `<R>` because the former is 0 arguments and the latter is 1.

File: heapster-saw/doc/tutorial/tutorial.md
Comment (scuellar) [reply heapster-saw/doc/tutorial/tutorial.md:63]: What do you have in mind for this link? Do you mean a hyperlink to the README in the master branch in github?

The location would be different if the tutorial is being read online or locally. That's why I tend to use local commands when I can.

File: heapster-saw/doc/tutorial/tutorial.md
Comment (scuellar) [reply heapster-saw/doc/tutorial/tutorial.md:349]: If it makes sense, I'll add it at the beginning of the Heapster section, since installation and running Saw makes more sense from the top level directory.

File: heapster-saw/doc/tutorial/tutorial.md
Comment (scuellar) [reply heapster-saw/doc/tutorial/tutorial.md:502]: Yes this makes sense. Since we are also adding a simpler example, I think I can directly show the type and explain it.

**Comment by scuellar** [comment]
This PR also fixes alignment in `CompMExtra.v` to match the standard linter. This is not mentioned in the commits, but it's obvious in the diff. 

The only substantive changes to `CompMExtra.v` are in the new module `Notation`

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:593]:
```suggestion
of the function to typecheck and its permission type. The command then
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:601]:
```suggestion
experiment with mistyping. Try running the following command
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:631]:
```suggestion
directory. You should see a handful of auto-generated imports and four
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:645]:
```suggestion
the `add` function and a unit `()`, representing the end of
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:704]:
```suggestion
Our first proof will claim that the function `add` produces no
errors. More specifically, it says that for all inputs (that's what
`refinesFun` quantifies) `add` always refines to `noErrorsSpec`. That
is, it returns a pure value.
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:762]:
```suggestion
As before, the ghost environment is omitted and both sides of the
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:770]:
```suggestion
argument describes the pointer offset, here `0`. Finally, the third
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:787]:
```suggestion
specification of `add`, `add_mistyped` and `incr_ptr`. As you can see
```

File: heapster-saw/examples/tutorial_c.c
Comment (RyanGlScott) [inline heapster-saw/examples/tutorial_c.c:12]:
```suggestion
// Struct that represents the three coordinates for a 3D vector
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:811]:
```suggestion
// Struct that represents the three coordinates for a 3D vector
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:844]: Wait, why does the function change the memory? I wouldn't have expected that since it's using the `R` modality.

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:919]:
```suggestion
We will briefly explore arrays, which have slightly more interesting
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:925]:
```suggestion
We will consider the function `zero_array` which zeroes out all the
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:939]:
```suggestion
length of the array and equal to the second argument (see the code in `arrays.saw`)
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:950]:
```suggestion
is the offset into the array. We represent that with a new ghost
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:962]:
```suggestion
adds dynamic checks on the extracted code which we will see in the Coq
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:965]:
```suggestion
Let's go to `arrays_gen.v` (which has already been generated for you)
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:969]:
```suggestion
out-of-bounds array accesses and unrepresentable indices (i.e. index
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:1009]:
```suggestion
condition, the second is added by Heapster to ensure the ghost
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:1020]:
```suggestion
If all the checks pass, then the function simply calls
itself recursively, with the next index and array with a new entry zeroed out.
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:1062]:
```suggestion
handle refinements with and without recursion (respectively). The rest
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:1200]:
```suggestion
determines if the list is readable or writable, just like 3D
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:1213]:
```suggestion
The definition shows the diferent cases for a list, separated by a
comma. In the first case, a `List64` can be a null pointer, expressed
with the type `eq(llvmword(0))`. In the second case, a list is a
pointer where offset `0` is the head, an `Int64`, and offset `8` is
the tail, is recursively a `List64<rw>`. In the later case,
both elements are tagged with `rw`, describing if they are readable or
writable, as determined by the argument to `List64<rw>`.
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:1217]:
```suggestion
`heapster_define_recursive_perm` command. Here is how to use the
```

File: heapster-saw/doc/tutorial/tutorial.md
Comment (RyanGlScott) [inline heapster-saw/doc/tutorial/tutorial.md:1232]:
```suggestion
that the permission applies to. The fifth argument is its permission
```

**Review by RyanGlScott — COMMENTED** [review]
I'm learning Heapster myself, so I found this to be quite helpful. I jotted down some minor spelling suggestions and general questions that I had while reading this.

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [reply heapster-saw/doc/tutorial/tutorial.md:12]: Agreed, please use SAW and SAWScript

**Review by eddywestbrook — CHANGES_REQUESTED** [review]
This looks really good, Santiago, thanks! Please do address the typos pointed out by Ryan before merging this.

File: heapster-saw/doc/tutorial/tutorial.md
Comment (eddywestbrook) [reply heapster-saw/doc/tutorial/tutorial.md:844]: Actually, yes, `norm_vector` should just have `arg0:true` as its output permission. The convention is that any copyable permissions in the input, such as `eq` or read-only pointer permissions, do not need to get returned in the output.

**Review by eddywestbrook — APPROVED** [review]
Thanks for writing this tutorial, Santiago!

**Pull request merged**
